src/Guide/State.hs

```diff
@@ -860,6 +875,38 @@ logoutUserGlobally key = do
 getAdminUsers :: Acid.Query GlobalState [User]
 getAdminUsers = filter (^. userIsAdmin) . toList <$> view users
 
+-- | Add external credentials to a user account.
+addCreds :: User -> Creds -> Acid.Update GlobalState Bool
+addCreds user userCreds = do
+  allCreds <- concat . toList <$> use creds
+  if canAddCreds userCreds allCreds
+  then do
+    creds %= M.insertWith (++) (user ^. userID) [userCreds]
+    return True
+  else
+    return False
+
+-- | Remove external credentials to a user account.
+removeCreds :: User -> Creds -> Acid.Update GlobalState ()
+removeCreds user userCreds = do
+  creds %= M.adjust (filter filterCred) (user ^. userID)
+  where
+    filterCred cred |    cred ^. credsProvider == userCreds ^. credsProvider
+                      && cred ^. credsId       == userCreds ^. credsId
+                      = False
+                    | otherwise = True
+
+-- | Get user from creds
+loginUserCreds :: Creds -> Acid.Query GlobalState (Maybe User)
+loginUserCreds userCreds = do
+  matches <- filter (\(_, cs) -> any matchCred cs) . M.toList <$> view creds
+  case matches of
+    [(key, _)] -> view (users . at key)
+    _ -> return Nothing
```

Comment:
Am I right that if there is more than one match, it's a bug? If yes, we should `error` out so that later we'll see in server logs that there was a bug.